In [1]:
#import libraries
import pandas as pd

In [2]:
#read combinedDonorDetailsData.csv

donor_df = pd.read_csv('combinedDonorDetailsData.csv')

In [3]:
donor_df.head()

donation_year                                         donor_name  \
0           2019  THE MACDONALD STEWART FOUNDATION - LA FONDATIO...   
1           2018  THE MACDONALD STEWART FOUNDATION - LA FONDATIO...   
2           2021                                CANADAHELPS CANADON   
3           2021                               KIND CANADA GÉNÉREUX   
4           2021  UNITED WAY EAST ONTARIO / CENTRAIDE EST DE L'O...   

  donor_city donor_province donation_amount donation_in_kind  \
0   MONTREAL             QC          14,000                0   
1   MONTREAL             QC           5,000                0   
2    TORONTO             ON           7,286                0   
3     OTTAWA             ON             500                0   
4     OTTAWA             ON             360                0   

                                        charity_link  \
0  https://www.charitydata.ca/charity/78th-fraser...   
1  https://www.charitydata.ca/charity/78th-fraser...   
2  https://www.charitydata.ca/charity/a-company-o...   
3  https://www.charitydata.ca/charity/a-company-o...   
4  https://www.charitydata.ca/charity/a-company-o...   

                                        charity_name  \
0  78TH FRASER HIGHLANDERS - FORT VICTORIA HISTOR...   
1  78TH FRASER HIGHLANDERS - FORT VICTORIA HISTOR...   
2                     A COMPANY OF FOOLS THEATRE INC   
3                     A COMPANY OF FOOLS THEATRE INC   
4                     A COMPANY OF FOOLS THEATRE INC   

               charity_type  \
0  Advancement of Education   
1  Advancement of Education   
2  Advancement of Education   
3  Advancement of Education   
4  Advancement of Education   

                                charity_sub_category      charity_description  \
0  Educational organizations not elsewhere catego...  Charitable Organization   
1  Educational organizations not elsewhere catego...  Charitable Organization   
2  Education in the arts / Theatre / film / drama...  Charitable Organization   
3  Education in the arts / Theatre / film / drama...  Charitable Organization   
4  Education in the arts / Theatre / film / drama...  Charitable Organization   

  charity_registration_number          charity_website charity_country  \
0             845496207RR0001  WWW.78FRASERVICTORIA.CA              CA   
1             845496207RR0001  WWW.78FRASERVICTORIA.CA              CA   
2             879738623RR0001             WWW.FOOLS.CA              CA   
3             879738623RR0001             WWW.FOOLS.CA              CA   
4             879738623RR0001             WWW.FOOLS.CA              CA   

   charity_province charity_city  charity_employee_range  
0  British Columbia       SIDNEY                       0  
1  British Columbia       SIDNEY                       0  
2           Ontario       OTTAWA                       0  
3           Ontario       OTTAWA                       0  
4           Ontario       OTTAWA                       0

In [4]:
donor_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83528 entries, 0 to 83527
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   donation_year                83528 non-null  int64 
 1   donor_name                   83527 non-null  object
 2   donor_city                   83418 non-null  object
 3   donor_province               83378 non-null  object
 4   donation_amount              83528 non-null  object
 5   donation_in_kind             83528 non-null  object
 6   charity_link                 83528 non-null  object
 7   charity_name                 83528 non-null  object
 8   charity_type                 83528 non-null  object
 9   charity_sub_category         83528 non-null  object
 10  charity_description          83528 non-null  object
 11  charity_registration_number  83528 non-null  object
 12  charity_website              59791 non-null  object
 13  charity_country              83

`donation_amount` and `donation_in_kind` are float values. Hence, change the datatype of these columns

In [6]:
#remove coma from the string values of these columns
donor_df['donation_amount'] = donor_df['donation_amount'] .apply(lambda x: x.replace(',', ''))
donor_df['donation_in_kind'] = donor_df['donation_in_kind'] .apply(lambda x: x.replace(',', ''))

#cast to float
donor_df['donation_amount'] = donor_df['donation_amount'].astype(float)
donor_df['donation_in_kind'] = donor_df['donation_in_kind'].astype(float)

Check if `donation_amount` and `donation_in_kind` have 0 as value. Remove the records that have 0 for both these columns

In [10]:
check_zero_df = donor_df.loc[(donor_df['donation_amount'] == 0) & (donor_df['donation_in_kind'] == 0)]

In [15]:
len(check_zero_df)

1205

There 1205 records with value 0 for both the columns. Hence, they should be removed.

In [20]:
donor_df = donor_df.loc[(donor_df['donation_amount'] != 0) | (donor_df['donation_in_kind'] != 0)]

Check if donation_amount and donation_in_kind have negative value. Remove those records

In [21]:
check_negative_df = donor_df.loc[(donor_df['donation_amount'] < 0) | (donor_df['donation_in_kind'] < 0)]

In [22]:
len(check_negative_df)

50

There are 50 records that have negative values for donation_amount and donation_in_kind. Remove them

In [26]:
donor_df = donor_df.loc[(donor_df['donation_amount'] >= 0) & (donor_df['donation_in_kind'] >= 0)]

In [27]:
distinct_provinces_donor = donor_df['donor_province'].unique() 

In [28]:
distinct_provinces_charity = donor_df['charity_province'].unique() 

In [29]:
distinct_provinces_donor

array(['QC', 'ON', 'AB', nan, 'BC', 'NL', 'MB', 'NS', 'NB', 'PE', 'SK',
       'CO', 'FL', 'YT', 'NT', 'CT', 'NY', 'CA'], dtype=object)

In [31]:
len(distinct_provinces_donor)

18

In [30]:
distinct_provinces_charity

array(['British Columbia', 'Ontario', 'Quebec',
       'Newfoundland and Labrador', 'Prince Edward Island', 'Alberta',
       'Manitoba', 'Nova Scotia', 'Saskatchewan', 'New Brunswick',
       'Northwest Territories', 'Yukon', 'Nunavut'], dtype=object)

In [32]:
len(distinct_provinces_charity)

13

In [33]:
province_codes = {
    "British Columbia": "BC",
    "Ontario": "ON",
    "Quebec": "QC",
    "Newfoundland and Labrador": "NL",
    "Prince Edward Island": "PE",
    "Alberta": "AB",
    "Manitoba": "MB",
    "Nova Scotia": "NS",
    "Saskatchewan": "SK",
    "New Brunswick": "NB",
    "Northwest Territories": "NT",
    "Yukon": "YT",
    "Nunavut": "NU"
}

In [37]:
donor_df['charity_province'] = donor_df['charity_province'].map(province_codes)

In [38]:
donor_df['charity_province'].unique()

array(['BC', 'ON', 'QC', 'NL', 'PE', 'AB', 'MB', 'NS', 'SK', 'NB', 'NT',
       'YT', 'NU'], dtype=object)

In [40]:
len(donor_df)

82273

In [41]:
donor_df.to_csv('cleanedDonorDetails.csv', index=False)